python/pyarrow/src/arrow/python/udf.cc

```diff
@@ -105,21 +158,117 @@ Status RegisterScalarFunction(PyObject* user_function, ScalarUdfWrapperCallback
+      std::make_shared<OwnedRefNoGIL>(user_function), wrapper, options.output_type);
   compute::ScalarKernel kernel(
       compute::KernelSignature::Make(std::move(input_types), std::move(output_type),
                                      options.arity.is_varargs),
-      PythonUdfExec);
+      PythonUdfExec, kernel_init);
   kernel.data = std::move(udf_data);
 
   kernel.mem_allocation = compute::MemAllocation::NO_PREALLOCATE;
   kernel.null_handling = compute::NullHandling::COMPUTED_NO_PREALLOCATE;
   RETURN_NOT_OK(scalar_func->AddKernel(std::move(kernel)));
-  auto registry = compute::GetFunctionRegistry();
+  if (registry == NULLPTR) {
+    registry = compute::GetFunctionRegistry();
+  }
   RETURN_NOT_OK(registry->AddFunction(std::move(scalar_func)));
   return Status::OK();
 }
 
-}  // namespace py
+}  // namespace
+
+Status RegisterScalarFunction(PyObject* user_function, ScalarUdfWrapperCallback wrapper,
+                              const ScalarUdfOptions& options,
+                              compute::FunctionRegistry* registry) {
+  return RegisterScalarLikeFunction(
+      user_function,
+      PythonScalarUdfKernelInit{std::make_shared<OwnedRefNoGIL>(user_function)}, wrapper,
+      options, registry);
+}
+
+Status RegisterTabularFunction(PyObject* user_function, ScalarUdfWrapperCallback wrapper,
+                               const ScalarUdfOptions& options,
+                               compute::FunctionRegistry* registry) {
+  if (options.arity.num_args != 0 || options.arity.is_varargs) {
+    return Status::NotImplemented("tabular function of non-null arity");
+  }
+  if (options.output_type->id() != Type::type::STRUCT) {
+    return Status::Invalid("tabular function with non-struct output");
+  }
+  return RegisterScalarLikeFunction(
+      user_function,
+      PythonTableUdfKernelInit{std::make_shared<OwnedRefNoGIL>(user_function), wrapper},
+      wrapper, options, registry);
+}
 
+namespace  {
+
+Result<std::shared_ptr<RecordBatch>> RecordBatchFromArray(
+    std::shared_ptr<Schema> schema, std::shared_ptr<Array> array) {
+  auto& data = const_cast<std::shared_ptr<ArrayData>&>(array->data());
+  if (data->child_data.size() != static_cast<size_t>(schema->num_fields())) {
+    return Status::Invalid("UDF result with shape not conforming to schema");
+  }
+  return RecordBatch::Make(std::move(schema), data->length, std::move(data->child_data));
+}
+
+}  // namespace
+
+Result<std::shared_ptr<RecordBatchReader>> CallTabularFunction(
+    const std::string& func_name, const std::vector<Datum>& args,
+    compute::FunctionRegistry* registry) {
+  if (args.size() != 0) {
+    return Status::NotImplemented("non-empty arguments to tabular function");
+  }
+  if (registry == NULLPTR) {
+    registry = compute::GetFunctionRegistry();
+  }
+  ARROW_ASSIGN_OR_RAISE(auto func, registry->GetFunction(func_name));
+  if (func->kind() != compute::Function::SCALAR) {
```

Comment:
I think we can tackle this in a future PR but I think we will eventually want to introduce a new function kind for tabular functions.  Or, if we have to pick one of the two options we have today, I think vector would be more accurate.  Since we have no inputs I think we might get away with this for now but if we had inputs I would expect the length of the array returned from a scalar function to match the length of the incoming inputs and I don't think this will be true here.

Reply:
Sounds reasonable to me.

Reply:
Thinking about this some more, since "tabular"characterizes the output of a function whereas "scalar"/"vector" characterizes its unit of operation, a tabular function may be tabular-scalar or tabular-vector. So, I think the right way to capture a function being tabular is something to discuss (separately).

Reply:
What would an example of a tabular-scalar function be?  Would that be a function that returns one row for each input row but might return many columns?

Reply:
Exactly.

Reply:
This particular line will still need to change at some point though right?  Right now you're asserting this must be scalar but the tabular functions you are interested in (generators) are not scalar.

Reply:
Yes, this will need to change late when we distinguish between a tabular-attribute and a scalar-attribute of a function.